src/utils/sqs_consumer.ts

```diff
@@ -0,0 +1,102 @@
+import { SQS } from 'aws-sdk';
+
+import { logger } from '../logger';
+
+export class SqsConsumer {
+    private readonly _sqs: SQS;
+    private readonly _queueUrl: string;
+    private readonly _beforeHandle?: () => Promise<boolean>;
+    private readonly _handleMessage: (message: SQS.Types.Message) => Promise<any>;
+    private readonly _afterHandle?: () => Promise<void>;
+    private _isConsuming: boolean;
+
+    constructor(params: {
+        sqs: SQS;
+        queueUrl: string;
+        beforeHandle?: () => Promise<boolean>;
+        handleMessage: (message: SQS.Types.Message) => Promise<any>;
+        afterHandle?: () => Promise<void>;
+    }) {
+        this._sqs = params.sqs;
+        this._queueUrl = params.queueUrl;
+        this._beforeHandle = params.beforeHandle;
+        this._handleMessage = params.handleMessage;
+        this._afterHandle = params.afterHandle;
+        this._isConsuming = false;
+    }
+
+    public stop(): void {
+        this._isConsuming = false;
+    }
+
+    public async consumeAsync(): Promise<void> {
+        this._isConsuming = true;
+        while (this._isConsuming) {
+            await this._consumeOnceAsync();
+        }
+    }
+
+    private async _consumeOnceAsync(): Promise<void> {
+        // Run the before hook
+        if (this._beforeHandle) {
+            let beforeCheck;
+            try {
+                beforeCheck = await this._beforeHandle();
+            } catch (e) {
+                logger.warn(e, 'Error encountered in the preHandle check');
+                return;
+            }
+
+            if (!beforeCheck) {
+                logger.warn('before validation failed');
+                return;
+            }
+        }
+
+        // Receive message
+        const response = await this._sqs
+            .receiveMessage({
+                MaxNumberOfMessages: 1,
+                WaitTimeSeconds: 20, // long polling
+                QueueUrl: this._queueUrl,
+            })
+            .promise();
+
+        // No message
+        if (response === undefined || response.Messages?.length !== 1) {
+            return;
```

Comment:
Can we add some logging here justto be verbose? `Fetched xxx messages..`

Reply:
I thought about it, but we'll be logging quite a bit - every 20 seconds, each worker will emit a log that says "Fetched 0 messages...". Thoughts on the value of that?

Reply:
Good question, maybe it's not needed? I was thinking it could give us a sense of if the worker got stuck somehow? What do you think?

Reply:
Hmm... if we want to get a sense for if the worker is trying to process something, we could add a request at the top of the workflow. I could see that being useful as a sanity check.

That said, I think we can add some logging in our worker liveness check and worker startup, so I imagine we'll have logging for that when we integrate that code.